Generate Serialization for FragmentedSharedBuffer

NKRosario opened the pull request:
#### 038987a9d4956dcfc6c850f53992645e4f6f4bd1
<pre>
Generate Serialization for FragmentedSharedBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=268859">https://bugs.webkit.org/show_bug.cgi?id=268859</a>
<a href="https://rdar.apple.com/122423871">rdar://122423871</a>

Reviewed by Chris Dumez.

* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::fromIPCData):
(WebCore::FragmentedSharedBuffer::toIPCData const):
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::useUnixDomainSockets): Deleted.
(IPC::ArgumentCoder&lt;WebCore::FragmentedSharedBuffer&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::FragmentedSharedBuffer&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/274563@main">https://commits.webkit.org/274563@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5287cbe26bd17af3c79b98240bdacd016b589959

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18424 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41979 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35345 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41751 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21301 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15753 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40019 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15522 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13485 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35116 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43257 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35816 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35446 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39247 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11757 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15863 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8823 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +68,23 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+    if (!handle.has_value())
+        return std::nullopt;
+
+    RefPtr<SharedMemory> sharedMemoryBuffer = SharedMemory::map(WTFMove(handle.value()), SharedMemory::Protection::ReadOnly);
```
`RefPtr<SharedMemory>` -> `RefPtr` suffices

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +139,15 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+#if PLATFORM(COCOA)
+std::optional<SharedMemoryHandle> FragmentedSharedBuffer::toSharedMemoryHandle() const
+{
+    RefPtr<SharedMemory> sharedMemoryBuffer = SharedMemory::copyBuffer(*this);
```
`RefPtr sharedMemoryBuffer`

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +139,15 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+std::optional<SharedMemoryHandle> FragmentedSharedBuffer::toSharedMemoryHandle() const
+{
+    RefPtr<SharedMemory> sharedMemoryBuffer = SharedMemory::copyBuffer(*this);
+    std::optional<SharedMemoryHandle> handle = sharedMemoryBuffer->createHandle(SharedMemory::Protection::ReadOnly);
```
No need for a local variable here, just:
`return sharedMemoryBuffer->createHandle(SharedMemory::Protection::ReadOnly)`

cdumez commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -26,6 +26,8 @@
 
 #pragma once
 
+#include "SharedMemory.h"
```
Would a forward declaration suffice? You included this in both the header and the cpp.

NKRosario replied:
just a forward declaration wasn't working. I removed it from the cpp but left it in the header

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -7749,3 +7749,10 @@ class WebCore::SerializedPlatformDataCueValue {
+#if PLATFORM(COCOA)
+header: <WebCore/SharedBuffer.h>
+[RefCounted, CustomHeader] class WebCore::FragmentedSharedBuffer {
+    std::optional<WebCore::SharedMemoryHandle> toSharedMemoryHandle();
```
Now we're creating a shared memory even if `bufferSize < minimumPageSize`, which seems like a performance regression compared to the previous code.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +69,33 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
 }
 
+#if PLATFORM(COCOA)
+std::optional<Ref<FragmentedSharedBuffer>> FragmentedSharedBuffer::fromIPCData(FragmentedSharedBuffer::FragmentedSharedBufferData&& fragmentedSharedBufferData)
```
`FragmentedSharedBuffer::` is unnecessary.

NKRosario replied:
it wasn't finding it without `FragmentedSharedBuffer::`

cdumez replied:
I am talking about the parameter type -> `FragmentedSharedBuffer::FragmentedSharedBufferData&&` should be `FragmentedSharedBufferData&&`

NKRosario replied:
I know, it wasn't finding it. It seems to be working now with IPCData though

cdumez commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -144,13 +145,18 @@ class DataSegment : public ThreadSafeRefCounted<DataSegment> {
 
 class FragmentedSharedBuffer : public ThreadSafeRefCounted<FragmentedSharedBuffer> {
 public:
+    using FragmentedSharedBufferData = std::variant<std::optional<WebCore::SharedMemoryHandle>, std::span<const uint8_t>>;
```
Can we call this `IPCData`, this is what we used elsewhere.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +150,27 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+    if (!bufferSize)
+        return std::span<const uint8_t> { };
+
+    if (bufferSize < minimumPageSize) {
```
Isn't it as easy as adding a `useUnixDomainSockets() ||` here to support non-COCOA ports now?

NKRosario replied:
yes, I can add that back

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +150,27 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+        if (m_segments.size() == 1)
+            return std::span { m_segments[0].segment->data(), m_segments[0].segment->size() };
+        auto combinedData = combineSegmentsData(m_segments, m_size);
+        return combinedData.span();
```
Sorry, I think I gave you incorrect code here. Now that I think of it, it is probably not safe to return a span<> to a local Vector. The Vector goes out of scope after returning and the span points to bad data. We have to write this differently somehow. I'll think about it.

cdumez replied:
I think we could return a `Vector<std::span<const uint8_t>>` instead. Then on sender side, you iterate over each segment and append to that vector. This is more efficient since we don't need to copy the bytes. Also, we no longer need a local Vector so no lifetime issue.

On recipient side, you call `append()` already anyway so you can append() the spans one by one. WDYT?

NKRosario replied:
Do we want `Vector<std::span<const uint8_t>>` or just `Vector<uint8_t>`?

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +78,32 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+    return WTF::switchOn(WTFMove(ipcData), [](Vector<std::span<const uint8_t>>&& data) -> std::optional<Ref<FragmentedSharedBuffer>> {
+        size_t size = sizeof(data[0]) * data.size();
+        if (useUnixDomainSockets() || size < minimumPageSize) {
+            Ref sharedMemoryBuffer = SharedBuffer::create();
```
This needs to construct a `FragmentedSharedBuffer`, not a `SharedBuffer` in the case there is more than one segment. We may want to optimize the case where there is a single vector and return `SharedBuffer::create(firstSegment)` in that case though.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +78,32 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+std::optional<Ref<FragmentedSharedBuffer>> FragmentedSharedBuffer::fromIPCData(IPCData&& ipcData)
+{
+    return WTF::switchOn(WTFMove(ipcData), [](Vector<std::span<const uint8_t>>&& data) -> std::optional<Ref<FragmentedSharedBuffer>> {
+        size_t size = sizeof(data[0]) * data.size();
```
Accessing `data[0]` looks dangerous here, especially considering the data comes from IPC and is thus untrusted.
Als, sizeof(data[0]) is always 1 anyway since it's a uint8_t.

cdumez replied:
Also, if you truly want to get the size of the data (and I think you do for proper validation), you'll need to iterate the vector and sum the size of each span it contains.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +158,26 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+        if (m_segments.isEmpty())
+            return { };
+        WTF::Vector<std::span<const uint8_t>> data;
+        data.reserveInitialCapacity(m_segments.size() * sizeof(std::span<const uint8_t>));
```
No `* sizeof(std::span<const uint8_t>)` here.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +158,26 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+            return { };
+        WTF::Vector<std::span<const uint8_t>> data;
+        data.reserveInitialCapacity(m_segments.size() * sizeof(std::span<const uint8_t>));
+        for (auto segment : m_segments)
```
`auto&`

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +158,26 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+        WTF::Vector<std::span<const uint8_t>> data;
+        data.reserveInitialCapacity(m_segments.size() * sizeof(std::span<const uint8_t>));
+        for (auto segment : m_segments)
+            data.append(std::span { segment.segment->data(), segment.segment->size() });
```
The modern (and more efficient) WebKit way to write this is actually:
```
return WTF::map(m_segments, [](auto& segment) {
    return std::span { segment.segment->data(), segment.segment->size() };
});
```

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -35,6 +35,17 @@
 
+static constexpr size_t minimumPageSize = 4096;
+
+constexpr bool useUnixDomainSockets()
```
Maybe this can be a static global constexpr boolean variable instead. Not much point in have a function here I think (I know you just copied the code over).

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +158,26 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+{
+    size_t bufferSize = size();
+    if (!bufferSize)
+        return { };
```
May be nicer to return an empty Vector instead? Otherwise, it constructs a variant with a SharedMemoryHandle that is nullopt, which should not be valid. However, sending an empty buffer should be OK I think.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +158,26 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+        return { };
+
+    if (useUnixDomainSockets() || bufferSize < minimumPageSize) {
+        if (m_segments.isEmpty())
```
We probably don't need this check, I think the code below will do the right thing.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +78,32 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+        }
+        return std::nullopt;
+    }, [](std::optional<WebCore::SharedMemoryHandle>&& handle) -> std::optional<Ref<FragmentedSharedBuffer>> {
+        if (useUnixDomainSockets() || !handle.has_value() || (handle->size() && handle->size() < minimumPageSize))
```
If the encoder returned an empty Vector in the empty case, we wouldn't need `handle->size() && `

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +78,32 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+    }, [](std::optional<WebCore::SharedMemoryHandle>&& handle) -> std::optional<Ref<FragmentedSharedBuffer>> {
+        if (useUnixDomainSockets() || !handle.has_value() || (handle->size() && handle->size() < minimumPageSize))
+            return std::nullopt;
+        if (!handle->size())
```
I think the sender should return an empty Vector in the empty case, so this logic should probably move to the lambda that takes a Vector.

cdumez reviewed (changes requested):
EWS is red and the code needs some fixing.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +74,38 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+        if (data.empty())
+            return SharedBuffer::create();
+
+        size_t size = 0;
```
We should use `CheckedSize` and deal with potential overflows.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +74,38 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+            if (data.size() == 1)
+                return SharedBuffer::create(data[0]);
+            Ref sharedMemoryBuffer = FragmentedSharedBuffer::create();
+            for (auto& span : data)
```
I think `auto` is fine for span, no need for `auto&`. spans are super cheap to copy.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +160,24 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+        return WTF::Vector<std::span<const uint8_t>>();
+
+    if (useUnixDomainSockets || bufferSize < minimumPageSize) {
+        WTF::Vector<std::span<const uint8_t>> data;
```
no longer needed.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +160,24 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+
+    if (useUnixDomainSockets || bufferSize < minimumPageSize) {
+        WTF::Vector<std::span<const uint8_t>> data;
+        data.reserveInitialCapacity(m_segments.size());
```
no longer needed.

cdumez commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -26,6 +26,8 @@
 
 #pragma once
 
+#include "SharedMemory.h"
```
Are you really just we can't just do `class SharedMemoryHandle;` to forward-declare? :)

NKRosario replied:
I'll try it again

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -121,6 +160,24 @@ Ref<SharedBuffer> FragmentedSharedBuffer::makeContiguous() const
+{
+    size_t bufferSize = size();
+    if (!bufferSize)
+        return WTF::Vector<std::span<const uint8_t>>();
```
We probably don’t need this early return. I think the code below will already do the right thing.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +74,38 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
     return adoptRef(*new FragmentedSharedBuffer(WTFMove(provider)));
 }
 
+#if PLATFORM(COCOA)
```
You need to drop this to fix the Linux build.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -67,6 +74,38 @@ Ref<FragmentedSharedBuffer> FragmentedSharedBuffer::create(DataSegment::Provider
+        CheckedSize size = 0;
+        for (auto span : data)
+            size += span.size();
+
```
I think we want to deal with overflow here and early return:
```
if (size.hasOverflowed())
    return std::nullopt;
```

Otherwise, we'll crash when using the `size` below.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -35,6 +35,13 @@
 
+static constexpr size_t minimumPageSize = 4096;
+#if USE(UNIX_DOMAIN_SOCKETS)
+static constexpr bool useUnixDomainSockets = 1;
```
please use `true`

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -35,6 +35,13 @@
+#if USE(UNIX_DOMAIN_SOCKETS)
+static constexpr bool useUnixDomainSockets = 1;
+#else
+static constexpr bool useUnixDomainSockets = 0;
```
please use `false`

cdumez reviewed (approved):
LGTM

NKRosario committed:
Generate Serialization for FragmentedSharedBuffer
https://bugs.webkit.org/show_bug.cgi?id=268859
rdar://122423871

Reviewed by Chris Dumez.

* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::fromIPCData):
(WebCore::FragmentedSharedBuffer::toIPCData const):
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::useUnixDomainSockets): Deleted.
(IPC::ArgumentCoder<WebCore::FragmentedSharedBuffer>::encode): Deleted.
(IPC::ArgumentCoder<WebCore::FragmentedSharedBuffer>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/274563@main
